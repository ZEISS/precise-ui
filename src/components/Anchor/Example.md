The `Anchor` is a generic component that can be used for routing purposes. It is unstyled and waits to be themed.

```jsx
const { Anchor } = require('precise-ui');

<Anchor href="https://google.com" target="_blank">Google</Anchor>
```

In conjunction with the React router a `to` is supported.

```jsx
const { Anchor } = require('precise-ui');

<Anchor to="/foo">Navigate to foo</Anchor>
```

Or any other allowed HTML element via its tag name.

```jsx
const { Anchor } = require('precise-ui');

<Anchor tagName="button" to="/foo">Navigate to foo</Anchor>
```
