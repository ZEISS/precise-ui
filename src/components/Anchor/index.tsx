import * as React from 'react';
import { StandardProps } from '../../common';

export interface AnchorUrlLocator {
  pathname?: string;
  search?: string;
  state?: any;
  hash?: string;
  key?: string;
}

export interface AnchorProps extends StandardProps {
  /**
   * Sets the component as disabled.
   */
  disabled?: boolean;
  /**
   * Redirects to the given URL if clicked.
   */
  href?: string;
  /**
   * Redirects to the given internal URL if clicked.
   */
  to?: string | AnchorUrlLocator;
  /**
   * The tagName to use, e.g., a or button.
   * @default a
   */
  tagName?: string;
  /**
   * Used to notify once clicked.
   */
  onClick?(e: React.MouseEvent<HTMLElement>): void;
  /**
   * Sets the action link children.
   */
  children?: React.ReactNode | void;
  /**
   * Specifies where to display the linked URL.
   * Applies only in case of `a` tag
   * Default is `_self`
   * @default _self
   */
  target?: '_self' | '_blank' | '_parent' | '_top';
}

function getHref(link: string | AnchorUrlLocator | undefined) {
  if (link !== undefined) {
    if (typeof link !== 'string') {
      const { pathname, search, hash } = link;
      const parts: Array<string> = [];

      if (pathname) {
        parts.push(pathname);
      }

      if (search) {
        parts.push('?');
        parts.push(search);
      }

      if (hash) {
        parts.push('#');
        parts.push(hash);
      }

      return parts.join('');
    }

    return link;
  }

  return undefined;
}

export class Anchor extends React.PureComponent<AnchorProps> {
  static contextTypes = {
    // tslint:disable-next-line
    router: () => null,
  };

  private clicked = (e: React.MouseEvent<HTMLElement>) => {
    const { tagName = 'a', disabled = false, target = '_self', href, to, onClick } = this.props;

    if (!disabled) {
      if (typeof onClick === 'function') {
        onClick(e);

        if (e.defaultPrevented) {
          return;
        }
      }

      if (to !== undefined) {
        const ctx = this.context;
        const history = ctx && ctx.router && ctx.router.history;

        if (history) {
          history.push(to);
        }
      } else if (typeof href === 'string') {
        if (tagName !== 'a') {
          window.open(href, target);
        }

        return;
      } else if (tagName === 'button') {
        return;
      }
    }

    e.preventDefault();
  };

  render() {
    const { tagName = 'a', theme: _0, to, disabled, target, ...props } = this.props;

    if (tagName === 'a' && target) {
      (props as any).target = target;
    }

    return React.createElement(tagName, {
      href: getHref(to) || (props.onClick ? '#' : undefined),
      ...props,
      onClick: this.clicked,
    });
  }
}
