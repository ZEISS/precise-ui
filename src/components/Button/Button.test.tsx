import * as React from 'react';
import * as enzyme from 'enzyme';
import { Button } from './';
import { Anchor } from '../Anchor';

describe('<Button />', () => {
  it('should render an empty <Button> component', () => {
    const wrapper = enzyme.mount(<Button />);
    expect(wrapper.find(Anchor).length).toEqual(1);
  });

  it('should render a <Button> component with some content', () => {
    const wrapper = enzyme.mount(<Button>Content</Button>).mount();
    expect(wrapper.text()).toBe('Content');
  });

  it('should render an active <Button> with some content', () => {
    const wrapper = enzyme.shallow(<Button active>Content</Button>);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render an active disabled <Button> with some content', () => {
    const wrapper = enzyme.shallow(
      <Button disabled active>
        Content
      </Button>,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a <Button> with some content component', () => {
    const wrapper = enzyme.shallow(
      <Button href="foo">
        <div>Content</div>
      </Button>,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should execute the click handler if enabled', () => {
    let clicked = false;
    const clickHandler = () => {
      clicked = true;
    };
    const wrapper = enzyme.mount(<Button onClick={clickHandler}>Content</Button>);
    wrapper.simulate('click', {
      preventDefault() {},
    });
    expect(clicked).toBe(true);
  });

  it('should not execute the click handler if disabled', () => {
    let clicked = false;
    const clickHandler = () => {
      clicked = true;
    };
    const wrapper = enzyme.mount(
      <Button onClick={clickHandler} disabled>
        Content
      </Button>,
    );
    wrapper.simulate('click', {
      preventDefault() {},
    });
    expect(clicked).toBe(false);
  });

  it('should not fail if no history given', () => {
    let clicked = false;
    const clickHandler = () => {
      clicked = true;
    };
    const wrapper = enzyme.mount(
      <Button to="/foo" onClick={clickHandler}>
        Content
      </Button>,
    );
    wrapper.simulate('click', {
      preventDefault() {},
    });
    expect(clicked).toBe(true);
  });
});
