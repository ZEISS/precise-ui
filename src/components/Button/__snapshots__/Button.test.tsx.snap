// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render a <Button> with some content component 1`] = `
<Styled(Component)
  href="foo"
  tagName="button"
>
  <styled.div>
    <div>
      Content
    </div>
  </styled.div>
</Styled(Component)>
`;

exports[`<Button /> should render an active <Button> with some content 1`] = `
<Styled(Component)
  active={true}
  tagName="button"
>
  <styled.div>
    Content
  </styled.div>
</Styled(Component)>
`;

exports[`<Button /> should render an active disabled <Button> with some content 1`] = `
<Styled(Component)
  active={true}
  disabled={true}
  tagName="button"
>
  <styled.div>
    Content
  </styled.div>
</Styled(Component)>
`;
